drivers/adc/ad9656/ad9656.c

```diff
@@ -0,0 +1,248 @@
+	 * contains the bits 7-0 of the address of the register. Byte 2 contains
+	 * the data to be written.
+	 */
+	buf[0] = reg_addr >> 8;
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = reg_data;
+
+	return spi_write_and_read(dev->spi_desc, buf, 3);
+}
+
+/**
+ * @brief Launch the JESD204 test functionality for the receive path for the ad9656 chip
+ * @param dev - The device handler for the ad9656 chip
+ * @param test_mode - The type of test that is to be performed or OFF if the testing
+ * 					  process is to be stopped
+ * @return SUCCESS if the ad9656 chip could be successfully set for JESD204 link testing,
+ * 		   FAILURE otherwise
+ */
+int32_t ad9656_JESD204_test(struct ad9656_dev *dev,
+			    uint32_t test_mode)
+{
+	uint8_t format;
+	ad9656_spi_write(dev, AD9656_REG_ADC_TEST_MODE, test_mode);
+	if (test_mode == AD9656_TEST_OFF)
+		format = AD9656_FORMAT_OFFSET_BINARY;
+	else
+		format = AD9656_FORMAT_2S_COMPLEMENT;
+
+	return ad9656_spi_write(dev, AD9656_REG_OUTPUT_MODE, format);
+}
+
+/**
+ * @brief Launch the user input test functionality for the receive path for the ad9656 chip.
+ * 	      User input data is supplied on the outputs instead of real data captured on the ADC.
+ * @param dev - The device handler for the ad9656 chip
+ * @param test_mode - The type of test that is to be performed or OFF if the testing
+ * 					  process is to be stopped
+ * @param user_input_test_pattern - User input test pattern that is sent on the output instead of
+ * 					the ADC	data
+ * @return SUCCESS if the ad9656 chip could be successfully set for user input testing, FAILURE
+ * 		   otherwise
+ */
+int32_t ad9656_user_input_test(struct ad9656_dev *dev, uint32_t test_mode,
+			       struct ad9656_user_input_test_pattern user_input_test_pattern)
+{
+	uint8_t format;
+
+	/* write user input test pattern in the registers */
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_LSB,
+			 user_input_test_pattern.user_test_pattern1 & 0xFF);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_MSB,
+			 user_input_test_pattern.user_test_pattern1 >> 8);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_2_LSB,
+			 user_input_test_pattern.user_test_pattern2 & 0xFF);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_2_MSB,
+			 user_input_test_pattern.user_test_pattern2 >> 8);
+
+	/* determine the chip to output the user test pattern */
+	return ad9656_spi_write(dev, AD9656_REG_ADC_TEST_MODE, test_mode);
+}
+
+/**
+ * @brief Setup the working parameters of the ad9656 chip
+ * @param device - The device handler of the ad9656 chip
+ * @param init_param - Values for the working parameters of ad9656
+ * @return SUCCESS if device is ready for use, FAILURE otherwise
+ */
+int32_t ad9656_setup(struct ad9656_dev **device,
+		     const struct ad9656_init_param *init_param)
+{
+	uint8_t chip_id;
+	uint8_t pll_stat;
+	int32_t ret;
+	struct ad9656_dev *dev;
+
+	dev = (struct ad9656_dev *)malloc(sizeof(*dev));
```

Comment:
calloc would also set the memory to 0 to be sure that there is no residual values left over in memory.

Reply:
Done in v6